### ISSUE TITLE
Logging capabilities for C++ submodels

### ISSUE THREAD

**Issue by diregoblin** [multiscale/muscle3#49]
Muscle 2 has a family of C++ functions, called muscle::logger::info/debug/etc., which can be used to output information about the simulation into a log file, as well as into stdout, if the severity is high enough. 
It would be nice to have similar capabilities in Muscle 3, ideally also with a similar API, for ease of porting models from Muscle 2 to Muscle 3.

**Comment by LourensVeen** [comment]
Would having these as member functions an the Instance class work for you? MUSCLE 2 has state in the library, which can lead to some very hard to debug issues, so I'm trying to keep everything in MUSCLE 3 attached to Instance. But that would mean that you'd have to pass your Instance object to wherever you need to log, or make it (or pointer or reference to it) a global.

A free function could log into a local file or to stdout, but it couldn't log high-priority messages via the network to the Manager. The latter is nice to have in HPC situations where the log file may be in a node-local temp dir which you no longer have access to after a crash, because someone else has that node now. Having the error message in the Manager's log is really nice to have in that case. Also, that would make it match Python.

On the other hand, adding logging functionality to MUSCLE 3 that just writes to standard out or a local file wouldn't make sense, as there would be nothing MUSCLE 3-specific about it. It would be better to have the user use a logging library then, or make a log function themselves. Okay, so I think I have convinced myself that this can be added to MUSCLE 3, but only as member functions on the Instance object. Would that work?

**Comment by diregoblin** [comment]
Well, that would work, but it would require me to go through more or less the entire ISR3D code and replace Muscle2-style logging with the new one. Passing the Instance everywhere doesn't sound like a good solution, since quite a lot of logging is done from the functions pretty deep inside the code that are not directly Muscle-related (i.e. a file-reading function might write a log message and exit with an error if the file is missing).

Actually, now that I think about it, those functions really have no business even knowing they are running inside a Muscle application, so they definitely should not use Muscle logging capabilities... 
I'll change those to use stdout and/or stderr for logs while porting the project to Muscle 3.
Also, I think that what you suggested in the email, having "the script/runner/pilot job framework redirect that to a log file" might be a very good way to do it.

For HPC environments, those logs would have to be collected to the same directory as the general model output, which can indeed be tricky in case of crashes... Purely local files can also be very useful, though, at least for me. I do a lot of debug runs either locally or on a dedicated node, which doesn't have a risk of someone else occupying it.

The bottom line is, I guess, that I have to get rid of Muscle 2 logging in ISR3D and write everything to cout for the Muscle3 port.

**Comment by LourensVeen** [comment]
The thing is also that there is probably code out there with a mishmash of logging calls, printf's, std::cout's, and so on. Or you have a model that logs in one way, but which uses a library which logs in another way. Pretty much any way of logging can be configured to print to stdout/stderr though, and doing that at least puts everything in a single location in more-or-less the right order. So it would be nice for MUSCLE 3, when it gains the capability to start submodels, to be able to deal with that, and the only way to do that would be by collecting stdout/stderr.

You can do that to a shared file system (/home on many HPC systems), but if you have many instances then the sysadmin is not going to be happy with you. So it's better to log everything to a tempdir on the node, then collect what you need afterwards, except that if you have a crash, you end up having the information you need to debug marooned on a node you can't reach. Hence the compromise of writing the bulk of the code to stdout/stderr and have that probably end up on the node at least temporarily, but allowing sending of high-priority log output (warnings, errors) to the Manager over the network.

But right now on C++ you can't do the latter, so this is a valid issue and I'll add some functions, but on the Instance class. Then if you run into an issue and you want to debug, you can at least quickly add a global pointer to the main file where you have your instance, set it to point to your Instance after you've created it, and use it in any other file to write log messages. That's an ugly hack, but it may just get you the information you need and your thesis finished in time, in which case it's worth it.

For ISR3D, I'd recommend creating a simple logging function that does the print and calling that, then if you change your mind later you can change the function in one place, rather than having to change it everywhere again.

**Comment by LourensVeen** [comment]
Oh, and a note about error handling, the clean way of doing that is to throw an exception, rather than quit somewhere at random. Even if the exception is not caught and the program crashes via std::terminate, MUSCLE 3 will be shut down cleanly in this case. This will help avoid a situation where one submodel crashes and the others just sit there waiting for it to send them a message.

Even better is to catch that exception at the top level and call [Instance::error_shutdown](https://muscle3.readthedocs.io/en/latest/cpp_api.html#_CPPv4N9libmuscle4impl8Instance14error_shutdownERKNSt6stringE), which will send the message you pass to the Manager for central logging and then shut down MUSCLE 3 cleanly.

The next version of MUSCLE 3 should start the submodels via a pilot job framework, so it will notice if the submodel crashed and shut everything down anyway, but the above is the correct way to do it.

- [ ] Put the above in the manual!
